### PR TITLE
Feat/data replicate service

### PR DESCRIPTION
**Pull request by giangbui**
resolve #98 and #100 via #322 #323

### CODE REVIEW

File: scripts/google_replicate.py [hunk 1]
```diff
@@ -0,0 +1,268 @@
...
+        """
+      	# un-comment this line for generating testing data.
+        # This test data contains real uuids and hashes and can be used for replicating
+        # GDC data to google bucket
+        # from intergration_data_test import google_gen_test_data
+        # submitting_files = google_gen_test_data()
+	submitting_files = get_fileinfo_list_from_manifest(self.manifest_file)
+        for th in self.thread_list:
+            th.start()
+        semaphoreLock.acquire()
```
[the remark this answers is not in the document]
Comment (giangbui) [reply scripts/google_replicate.py]: We do need the lock here since when threads start, they will start to get item from the queue.

File: scripts/aws_replicate.py [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+                    if not check_object(s3, self.bucket, object_name):
+                        if turn == 0:
+                            logger.info('object {} is not existed'.format(object_name))
+                        continue
+                    etag = get_etag_aws_object(s3, target_bucket, object_name)
+                    if etag is None or (etag.lower() != fi.get('hash','').lower()):
+                        execstr = execstr + \
+                            " --include \"{}\"".format(object_name)
+                if execstr != baseCmd:
+                    os.system(execstr)
```
[the remark this answers is not in the document]
Comment (giangbui) [reply scripts/aws_replicate.py]: switched in 99f5fd5a2b6249d7b13f046b401fffbd59169d34

File: scripts/google_replicate.py [hunk 3]
```diff
@@ -0,0 +1,268 @@
...
+                totalDownloadedBytes = totalDownloadedBytes + 100*chunk_size
+                semaphoreLock.release()
+            if chunk:  # filter out keep-alive new chunks
+                s.write(chunk)
+
+
+def process_data(threadName, global_config, q):
+    while not exitFlag:
+        semaphoreLock.acquire()
+        if not q.empty():
```
[the remark this answers is not in the document]
Comment (giangbui) [reply scripts/google_replicate.py]: The thread needs a lock so that it can access the queue. Moreover, when the queue is empty the thread will be terminated

File: scripts/aws_replicate.py [hunk 4]
```diff
@@ -0,0 +1,194 @@
...
+            # The system will log all the file with failure copy in the next step
+            for turn in xrange(0,2):
+                execstr = baseCmd
+                for fi in files[index:index + number_copying_files]:
+                    object_name = "{}/{}".format(fi.get("fileid"), fi.get("filename"))
+                    if not check_object(s3, self.bucket, object_name):
+                        if turn == 0:
+                            logger.info('object {} is not existed'.format(object_name))
+                        continue
+                    etag = get_etag_aws_object(s3, target_bucket, object_name)
```
[the remark this answers is not in the document]
Comment (giangbui) [reply scripts/aws_replicate.py]: Nice suggestion.

File: scripts/aws_replicate.py [hunk 5]
```diff
@@ -0,0 +1,194 @@
...
+            number_copying_files = min(chunk_size, len(files) - index)
+
+            # According to AWS document, copying failure is very rare. In order to handle the failure case,
+            # the system runs the below code two times, checking for the etag is performed for each time.
+            # The system will log all the file with failure copy in the next step
+            for turn in xrange(0,2):
+                execstr = baseCmd
+                for fi in files[index:index + number_copying_files]:
+                    object_name = "{}/{}".format(fi.get("fileid"), fi.get("filename"))
+                    if not check_object(s3, self.bucket, object_name):
```
[the remark this answers is not in the document]
Comment (giangbui) [reply scripts/aws_replicate.py]: changed in 99f5fd5a2b6249d7b13f046b401fffbd59169d34

File: scripts/google_replicate.py [hunk 6]
```diff
@@ -0,0 +1,268 @@
...
+
+global totalDownloadedBytes
+totalDownloadedBytes = 0
+
+global aliveThreads
+aliveThreads = 0
+
+logger = get_logger("ReplicationThread")
+
+semaphoreLock = threading.Lock()
```
[the remark this answers is not in the document]
Comment (giangbui) [reply scripts/google_replicate.py]: fixed in 99f5fd5a2b6249d7b13f046b401fffbd59169d34

File: scripts/utils.py [hunk 7]
```diff
@@ -0,0 +1,123 @@
+from binascii import unhexlify, hexlify
+import base64
+import re
+import argparse
+
+def Base64ToHexHash(base64_hash):
+    return hexlify(base64.decodestring(base64_hash.strip('\n"\'')))
+
+#st = Base64ToHexHash('YXg/2Z/PYhMRgRXqxJBjpg==')
```
Comment (zflamig) [inline scripts/utils.py]: Whats this?

File: scripts/settings.py [hunk 8]
```diff
@@ -0,0 +1,24 @@
+import os
+import json
+
+# Signpost
```
Comment (zflamig) [inline scripts/settings.py]: Change comment?

File: scripts/google_replicate.py [hunk 9]
```diff
@@ -0,0 +1,193 @@
...
+        client(GSClient): gs storage client
+        response(httpResponse): http response
+        bucket_name(str): target google bucket name
+        chunk_size_download(int): chunk size in bytes from downling data file from GDC
+        blob_name(str): object name
+    Returns:
+        None
+    """
+
+    # keep tract the number of chunks uploaded
```
Comment (zflamig) [inline scripts/google_replicate.py]: s/tract/track/

Comment (giangbui) [reply scripts/google_replicate.py]: Fixed at 77f150cb19e61bffaf580ba1b5531739f02920ae

File: scripts/settings.py (hunk 8 above)
Comment (giangbui) [reply scripts/settings.py]: Fixed at 77f150cb19e61bffaf580ba1b5531739f02920ae

File: scripts/utils.py (hunk 7 above)
Comment (giangbui) [reply scripts/utils.py]: I have just removed that line at 77f150cb19e61bffaf580ba1b5531739f02920ae

File: README.md [hunk 10]
```diff
@@ -1 +1,8 @@
-# dcf-dataservice
\ No newline at end of file
+# dcf-dataservice
+Copy data buckets from GDC to AWS and Google buckets. The jobs need to be run inside a setup kubeneste cluster.
+
+Following the below steps to setup
+- Create a directory in `$vpc_name/apis_configs` and named as `dcf_dataservice`.
+- Put the `manifest` file, `creds.json`, `aws_creds_secret` and `gcloud-creds-secret` into `dcf_dataservice` folder. While `aws_creds_secret` contains AWS key `gcloud-creds-secret` contains google cloud service account.
+- Run `jobs/kube-script-setup.sh`.
```
Comment (frickjack) [inline README.md]: Add a section explaining how the AWS sync works (just uses AWS API to sync in-AWS buckets),
and another explaining how the Google sync works (deploys a data-flow job that copies data from GDC to GCP buckets)

File: README.md [hunk 11]
```diff
@@ -1 +1,18 @@
-# dcf-dataservice
\ No newline at end of file
+# dcf-dataservice
+Copy data buckets from GDC to AWS and Google buckets. The jobs need to be run inside a setup kubeneste cluster.
```
Comment (Avantol13) [inline README.md]:
```suggestion
 Copy data buckets from GDC to AWS and Google buckets. The jobs need to be run inside a setup kubernetes cluster.
```

File: scripts/aws_replicate.py [hunk 12]
```diff
@@ -0,0 +1,376 @@
...
+            threadName(str): thread name
+            global_config(dict): configuration dictionary
+            copied_objects(set): objects aready copied
+            q(Queue): queue containing the objects need to be processed
+        
+        Return:
+            None
+        """
+
+        while not EXIT_FLAG:
```
Comment (rudyardrichter) [inline scripts/aws_replicate.py]: I think this logic with the globals is messy for a number of reasons. `EXIT_FLAG` should be replaced with `thread.exit()` for sure, and you could change `run` to return the number of processed objects and aggregate these in the replicator's `run`

File: scripts/aws_replicate.py [hunk 13]
```diff
@@ -0,0 +1,376 @@
...
+
+logger = get_logger("AWSReplication")
+
+global EXIT_FLAG
+EXIT_FLAG = 0
+
+global TOTAL_PROCESSED_FILES
+TOTAL_PROCESSED_FILES = 0
+
+mutexLock = threading.Lock()
```
Comment (rudyardrichter) [inline scripts/aws_replicate.py]: just put this in the `run` that runs the threads, and pass in a reference to the threads?

File: README.md [hunk 14]
```diff
@@ -1 +1,18 @@
...
+
+GDC has two different data storages: Data center and AWS backup bucket which are synchronized continuously with each other. Every one or two months, GDC releases new/updated/deleted data in manifest files, the goal is to deploy two separated jobs which copy the data to AWS, GOOGLE cloud respectively.
+
+## GDC AWS bucket to DCF AWS buckets
+Simplely using AWS API to sync in-AWS buckets. To boost up the performance, we deploy multiple-thread aws s3 copying.
```
Comment (Avantol13) [inline README.md]:
```suggestion
 Simply using AWS API to sync in-AWS buckets. To boost up the performance, we deploy multiple-thread aws s3 copying.
```

File: scripts/aws_replicate.py [hunk 15]
```diff
@@ -0,0 +1,376 @@
...
+
+        index = index + chunk_size
+
+    global TOTAL_PROCESSED_FILES
+    mutexLock.acquire()
+    TOTAL_PROCESSED_FILES += len(files)
+    mutexLock.release()
+
+
+def process_data(threadName, source_bucket, global_config, copied_objects, q):
```
Comment (rudyardrichter) [inline scripts/aws_replicate.py]: I think this should just be inlined inside the thread class, and then it will also be fine to get the references to a mutex etc from the caller

File: scripts/aws_replicate.py [hunk 16]
```diff
@@ -0,0 +1,376 @@
...
+    Args:
+        threadName(str): thread name
+        source_bucket(str): source bucket
+        global_config(dict): configuration dictionary
+        copied_objects(set): objects aready copied
+        q(Queue): queue containing the objects need to be processed
+    """
+    while not EXIT_FLAG:
+        mutexLock.acquire()
+        if not q.empty():
```
Comment (rudyardrichter) [inline scripts/aws_replicate.py]: flip this:
```python
if q.empty():
    mutexLock.release()
    return
files = q.get()
...
```

File: README.md [hunk 17]
```diff
@@ -1 +1,18 @@
...
+
+## GDC data center to DCF GOOGLE buckets
+Deploy a data-flow job that copies data from GDC to GCP buckets. The `copy` transform of the data-flow pipeline streams objects from GDC data center to GOOGLE buckets.
+
+## How to run
+Following the below steps to setup
+- Create a directory in `$vpc_name/apis_configs` and named as `dcf_dataservice`.
+- Put the `creds.json`, `aws_creds_secret` and `gcloud-creds-secret` into `dcf_dataservice` folder. While `aws_creds_secret` contains AWS key `gcloud-creds-secret` contains google cloud service account.
```
Comment (Avantol13) [inline README.md]: Is this correct? The google creds are a service account key in JSON format?
```suggestion
 - Put the `creds.json`, `aws_creds_secret` and `gcloud-creds-secret` into `dcf_dataservice` folder. While `aws_creds_secret` contains AWS key `gcloud-creds-secret` contains Google Cloud service account key (JSON credentials).
```

File: scripts/aws_replicate.py [hunk 18]
```diff
@@ -0,0 +1,376 @@
...
+        logger.info("start to index {}".format(object_path))
+        if object_path not in copied_objects:
+            json_log[object_path] = {"copy_success": False, "index_success": False}
+        else:
+            try:
+                json_log[object_path] = {
+                    "copy_success": True,
+                    "index_success": update_indexd(fi),
+                }
+            except Exception as e:
```
Comment (rudyardrichter) [inline scripts/aws_replicate.py]: it would be nice to make these `Exception` more specific

File: README.md (hunk 10 above)
Comment (Avantol13) [reply README.md]: +1 diagrams would be nice too so we can easily show this to other parties without having to require reading a block of text

File: dataflow_pipeline.py [hunk 19]
```diff
@@ -0,0 +1,111 @@
...
+from scripts.google_replicate import exec_google_copy
+
+try:
+    unicode  # pylint: disable=unicode-builtin
+except NameError:
+    unicode = str
+
+FILE_HEADERS = ["id", "filename", "md5", "size", "state", "acl", "project_id"]
+
+# global_config ={"token_path": "./gdc-token.txt",
```
Comment (Avantol13) [inline dataflow_pipeline.py]: dead code, remove?

File: scripts/aws_replicate.py [hunk 20]
```diff
@@ -0,0 +1,376 @@
...
+                logger.error(e.message)
+                json_log[object_path] = {
+                    "copy_success": True,
+                    "index_success": False,
+                    "msg": e.message,
+                }
+
+
+def update_indexd(fi):
```
Comment (rudyardrichter) [inline scripts/aws_replicate.py]: there's a lot of duplication between this and the google one, could these be consolidated?

File: dev-requirements.txt [hunk 21]
```diff
@@ -0,0 +1,9 @@
+mock==1.0.1
+pytest-cov==2.5.1
+coverage==3.7.1
+boto3>=1.7.28
+#google-cloud==0.33.1
```
Comment (Avantol13) [inline dev-requirements.txt]: dead code?

File: dataflow_pipeline.py [hunk 22]
```diff
@@ -0,0 +1,111 @@
...
+from apache_beam.io import WriteToText
+from apache_beam.metrics import Metrics
+from apache_beam.metrics.metric import MetricsFilter
+from apache_beam.options.pipeline_options import PipelineOptions
+from apache_beam.options.pipeline_options import SetupOptions
+import json
+
+from scripts.google_replicate import exec_google_copy
+
+try:
```
Comment (Avantol13) [inline dataflow_pipeline.py]: why is this necessary? can you add some comments explaining this? I assume it's some Python 2/3 compatibility?

File: dataflow_pipeline.py [hunk 23]
```diff
@@ -0,0 +1,111 @@
...
+        help="Input file to process.",
+    )
+    parser.add_argument(
+        "--output",
+        dest="output",
+        required=True,
+        help="Output file to write results to.",
+    )
+    parser.add_argument(
+        "--global_config", dest="global_config", help="global configuration"
```
Comment (Avantol13) [inline dataflow_pipeline.py]: what format should this global configuration be in? What are you passing? A file name? can you add more details to the `help` section explaining these things?

File: scripts/aws_replicate.py [hunk 24]
```diff
@@ -0,0 +1,376 @@
...
+TOTAL_PROCESSED_FILES = 0
+
+mutexLock = threading.Lock()
+workQueue = Queue.Queue()
+
+indexclient = IndexClient(
+    INDEXD["host"],
+    INDEXD["version"],
+    (INDEXD["auth"]["username"], INDEXD["auth"]["password"]),
+)
```
Comment (rudyardrichter) [inline scripts/aws_replicate.py]: really just all of these globals—please let's find a way to improve the ownership model here so the caller classes/functions have these, instead of making them global, and pass them down to whatever needs to have it

File: scripts/aws_replicate.py [hunk 25]
```diff
@@ -0,0 +1,376 @@
...
+        self.thread_list = []
+        self.manifest_file = manifest_file
+        self.global_config = global_config
+        self.thread_list = []
+        self.thread_num = thread_num
+        self.json_log = {}
+        self.job_name = job_name
+
+        copied_objects = self.get_copied_objects()
+        for i in xrange(0, self.thread_num):
```
Comment (rudyardrichter) [inline scripts/aws_replicate.py]: this can be list comprehension

File: dev-requirements.txt [hunk 26]
```diff
@@ -0,0 +1,9 @@
...
+google-cloud-storage==1.6.0
+-e git+https://git@github.com/uc-cdis/cdislogging.git@0.0.2#egg=cdislogging
+-e git+https://git@github.com/uc-cdis/indexclient.git@1.5.6#egg=indexclient
```
Comment (Avantol13) [inline dev-requirements.txt]: Why are we also including `indexclient-1.5.2.zip` as a zip file? this seems like it'll be difficult to maintain. also, the dev-requirements have a different version `1.5.6`

File: jobs/google-bucket-replicate-job.yaml [hunk 27]
```diff
@@ -0,0 +1,48 @@
+# run with 
+# gen3 runjob google-bucket-replicate-job.yaml PROJECT google_project INPUT_BUCKET input_bucket LOG_BUCKET log_bucket
+# Ex. gen3 runjob dcf-dataservice/jobs/google-bucket-replicate-job.yaml PROJECT cdis-test-188416 INPUT_BUCKET data-flow-code LOG_BUCKET data-flow-code
```
Comment (Avantol13) [inline jobs/google-bucket-replicate-job.yaml]: Should the `INPUT_BUCKET`  be the same as the `LOG_BUCKET` in the example? both are `data-flow-code`

File: jobs/google-bucket-replicate-job.yaml [hunk 28]
```diff
@@ -0,0 +1,48 @@
...
+            mountPath: "/secrets/dcf_dataservice_settings"
+            subPath: "dcf_dataservice_settings"
+        command: ["/bin/bash" ]
+        args: 
+          - "-c"
+          - |
+            cat /secrets/dcf_dataservice_settings > ./scripts/settings.py
+            gcloud auth activate-service-account --key-file=/secrets/google_service_account_creds
+            export GOOGLE_APPLICATION_CREDENTIALS=/secrets/google_service_account_creds
+            python dataflow_pipeline.py  --runner DataflowRunner --project $PROJECT --job_name dcf-dataservice --staging_location gs://$LOG_BUCKET/staging --temp_location gs://$LOG_BUCKET/temp --output gs://$LOG_BUCKET/output --setup_file ./setup.py --input gs://$INPUT_BUCKET/input/manifest --extra_package indexclient-1.5.2.zip
```
Comment (Avantol13) [inline jobs/google-bucket-replicate-job.yaml]: what's the difference between staging and temp location? should document that somewhere here or in the readme

File: README.md [hunk 29]
```diff
@@ -1 +1,18 @@
...
+- Put the `manifest` file into `dcf_dataservice`for AWS sync. For GOOLE sync, put the file into `gs://INPUT_BUCKET//input/`. We also need to have `LOG_BUCKET` to store the log and the output outcome. Please see the yaml job file for more detail.
+- Run `jobs/kube-script-setup.sh`.
```
Comment (Avantol13) [inline README.md]: `kube-script-setup` seems like a weird name for this. the intention is that this will primarily be used to prep for data replication, right? So maybe we should have a more specific name like `kube-setup-replication` or something? just an idea, dunno what else this is/will be used for

File: scripts/aws_replicate.py [hunk 30]
```diff
@@ -0,0 +1,376 @@
...
+            thread_num(int): number of threads
+            job_name(str): copying|indexing
+            bucket(str): source bucket
+        
+        """
+        self.bucket = bucket
+        self.thread_list = []
+        self.manifest_file = manifest_file
+        self.global_config = global_config
+        self.thread_list = []
```
Comment (rudyardrichter) [inline scripts/aws_replicate.py]: duplicated

File: jobs/kube-setup-scripts.sh [hunk 31]
```diff
@@ -0,0 +1,40 @@
+#!/bin/bash
```
Comment (Avantol13) [inline jobs/kube-setup-scripts.sh]: this seems specific to data replication, maybe we should name this file something like `kube-setup-replication.sh`?

File: setup.py [hunk 32]
```diff
@@ -0,0 +1,28 @@
...
+]
+
+PACKAGE_NAME = "scripts"
+PACKAGE_VERSION = "0.0.1"
+setuptools.setup(
+    name=PACKAGE_NAME,
+    version=PACKAGE_VERSION,
+    description="required dependencies",
+    install_requires=REQUIRED_PACKAGES,
+    # dependency_links = [
```
Comment (Avantol13) [inline setup.py]: dead code, remove?

File: scripts/google_replicate.py [hunk 33]
```diff
@@ -0,0 +1,240 @@
...
+    """
+    check if blob object exists or not
+    Args:
+        bucket_name(str): the name of bucket
+        blob_name(str): the name of blob
+        fi(dict): file info dictionary
+    Returns:
+        bool: indicating value if the blob is exist or not
+    """
+    url = "https://www.googleapis.com/storage/v1/b/{}/o/{}".format(
```
Comment (Avantol13) [inline scripts/google_replicate.py]: can you not use the `storage` library to get this information? I know google's api's are sometimes weird, so it's entirely possible you have to use the REST API, I'm just curious

File: scripts/google_replicate.py [hunk 34]
```diff
@@ -0,0 +1,240 @@
...
+def exec_google_copy(fi, global_config):
+    """
+    copy a file to google bucket.
+    Args:
+        fi(dict): a dictionary of a copying file
+        global_config(dict): a configuration
+    Returns:
+        DataFlowLog
+    """
+    client = storage.Client()
```
Comment (Avantol13) [inline scripts/google_replicate.py]: so, this is a huge change, but probably all of this Google logic should actually live in the `cirrus` library, as that is what its intention was, to wrap all this Google stuff so we can reuse elsewhere.

It handles setting up authorized session and everything, but I'm beginning to think that long-term we should really migrate a *lot* of this code to `cirrus`...

File: scripts/aws_replicate.py [hunk 35]
```diff
@@ -0,0 +1,376 @@
...
+                mutexLock.release()
+
+    def write_log_file(self, s3):
+        """
+        write log file and push to s3 log bucket
+        """
+        filename = self.threadName + "_log.json"
+        with open(filename, "w") as outfile:
+            json.dump(self.json_log, outfile)
+        s3.upload_file(
```
Comment (rudyardrichter) [inline scripts/aws_replicate.py]: you can upload straight to s3 without the temporary local file by making a `boto.s3.key.Key` and setting its contents, or using StringIO/BytesIO. (unless it's desirable to have the logs save here as well)

File: scripts/google_replicate.py (hunk 34 above)
Comment (Avantol13) [reply scripts/google_replicate.py]: 1) to allow for a single place to handle our interface to Google's API should they change things
2) for modularized code that can be imported and used in other future software that may need to do some of these general actions like copying to a bucket

File: scripts/aws_replicate.py (hunk 25 above)
Comment (rudyardrichter) [reply scripts/aws_replicate.py]: scratch this; see suggestion for ThreadPool

File: scripts/aws_replicate.py [hunk 36]
```diff
@@ -0,0 +1,376 @@
...
+        for th in self.thread_list:
+            th.start()
+
+    def run(self,):
+        global TOTAL_PROCESSED_FILES
+
+        start = timeit.default_timer()
+        # Wait for queue to empty
+        while not workQueue.empty():
+            time.sleep(5)
```
Comment (rudyardrichter) [inline scripts/aws_replicate.py]: I think a `threading.Event` is likely to be more elegant; you can have a thread to watch for completion, and a thread to print timing/progress

File: scripts/google_resumable_upload.py [hunk 37]
```diff
@@ -0,0 +1,72 @@
+from google.auth.transport.requests import AuthorizedSession
+from google.resumable_media import requests, common
+import logging
+
+
+class GCSObjectStreamUpload(object):
```
Comment (Avantol13) [inline scripts/google_resumable_upload.py]: can you add some docstrings and explanation of what this class is? is this based on some Google-provided sample code?

File: tests/test_replicate.py [hunk 38]
```diff
@@ -0,0 +1,187 @@
...
+    fake = {
+        "id": "11111111111111111",
+        "filename": "abc.bam",
+        "size": 1,
+        "md5": "1223344543t34mt43tb43ofh",
+        "acl": "tgca",
+        "project_id": "TGCA",
+    }
+
+    return [
```
Comment (rudyardrichter) [inline tests/test_replicate.py]: `n * [fake]`

File: scripts/replicate.py [hunk 39]
```diff
@@ -0,0 +1,62 @@
...
+    redact_cmd.add_argument("--log_file", required=True)
+
+    return parser.parse_args()
+
+
+if __name__ == "__main__":
+    start = timeit.default_timer()
+
+    args = parse_arguments()
+    if args.action == "aws_replicate":
```
Comment (Avantol13) [inline scripts/replicate.py]: is this only used for aws replicating? sorta confusing that there's a `aws_replicate.py` file and a `replicate.py` file that takes an aws arg. would be nice to consolidate. I could be misunderstanding the different use-cases though

File: scripts/aws_replicate.py [hunk 40]
```diff
@@ -0,0 +1,376 @@
...
+    Returns:
+        None
+    """
+
+    config_chunk_size = global_config.get("chunk_size", 1)
+
+    index = 0
+
+    while index < len(files):
+        base_cmd = 'aws s3 cp s3://{} s3://{} --recursive --exclude "*"'.format(
```
Comment (rudyardrichter) [inline scripts/aws_replicate.py]: is there a particular reason to do it with the CLI? it seems like the boto3 S3 client would be fine for this

File: scripts/utils.py [hunk 41]
```diff
@@ -0,0 +1,178 @@
...
+        hash(str) or None
+    """
+    m = re.search("[0-9,a-f]{32}", text.lower())
+    if m:
+        return m.group(0)
+    else:
+        return None
+
+
+def get_headers(manifest_file):
```
Comment (Avantol13) [inline scripts/utils.py]: missing docstring

File: scripts/utils.py [hunk 42]
```diff
@@ -0,0 +1,178 @@
...
+            if len(content) <= 1:
+                return []
+            return content[0].replace("\r", "").replace("\n", "").split("\t")
+
+    except IOError:
+        print("File {} is not existed".format(manifest_file))
+    return []
+
+
+def get_fileinfo_list_from_s3_manifest(url_manifeat):
```
Comment (Avantol13) [inline scripts/utils.py]: missing docstring

File: scripts/utils.py [hunk 43]
```diff
@@ -0,0 +1,178 @@
...
+                l.append(dictionary)
+
+            return l, headers
+
+    except IOError:
+        print("File {} is not existed".format(manifest_file))
+    return l, []
+
+
+def split_manifest_file(manifest_file, file_nums=1):
```
Comment (Avantol13) [inline scripts/utils.py]: missing docstring

File: scripts/utils.py [hunk 44]
```diff
@@ -0,0 +1,178 @@
...
+                        row.get(headers[1]),
+                        row.get(headers[2]),
+                        row.get(headers[3]),
+                        row.get(headers[4]),
+                        row.get(headers[5]),
+                    )
+                )
+
+
+def write_fileinfo_list(filepath, files):
```
Comment (Avantol13) [inline scripts/utils.py]: missing docstring

File: tests/test_replicate.py [hunk 45]
```diff
@@ -0,0 +1,187 @@
...
+        fake,
+    ]
+
+
+@patch("google.cloud.storage.Client")
+@patch("scripts.google_replicate.check_blob_name_exists_and_match_md5_size")
+@patch("scripts.google_replicate.bucket_exists")
+def test_resumable_streaming_copy_called(
+    mock_bucket_exists, mock_check_blob, mock_client
+):
```
Comment (Avantol13) [inline tests/test_replicate.py]: I would add some docstrings for these tests to clarify exactly what you're testing

File: scripts/utils.py [hunk 46]
```diff
@@ -0,0 +1,178 @@
...
+            sub_rows = rows[
+                file_index * nrow_in_subfile : (file_index + 1) * nrow_in_subfile
+            ]
+        else:
+            sub_rows = rows[file_index * nrow_in_subfile :]
+
+        new_filename = "{}_{}".format(manifest_file, file_index)
+        with open(new_filename, "w") as writer:
+            writer.write(
+                "{}\t{}\t{}\t{}\t{}\t{}".format(
```
Comment (rudyardrichter) [inline scripts/utils.py]:
```python
"\t".join(headers[:5])
```

File: scripts/utils.py [hunk 47]
```diff
@@ -0,0 +1,178 @@
...
+
+def get_headers(manifest_file):
+    try:
+        with open(manifest_file, "r") as f:
+            content = f.readlines()
+            if len(content) <= 1:
+                return []
+            return content[0].replace("\r", "").replace("\n", "").split("\t")
+
+    except IOError:
```
Comment (rudyardrichter) [inline scripts/utils.py]: I think skip the IOError checks in these called functions, just put a check in the caller to make sure the file exists. then you don't have to have the separate IOError checks everywhere

**Comment by Avantol13** [comment]
oh, I also only checked the google stuff

File: scripts/aws_replicate.py [hunk 48]
```diff
@@ -0,0 +1,376 @@
...
+from utils import (
+    get_bucket_name,
+    get_fileinfo_list_from_manifest,
+    get_fileinfo_list_from_s3_manifest,
+    exec_files_grouping,
+)
+
+logger = get_logger("AWSReplication")
+
+global EXIT_FLAG
```
[the remark this answers is not in the document]
Comment (giangbui) [reply scripts/aws_replicate.py]: switch to thread pool in d47a36c889171c98bbe27a87e89fb6d55d356f0f

File: scripts/aws_replicate.py (hunk 20 above)
Comment (giangbui) [reply scripts/aws_replicate.py]: Fixed in d47a36c889171c98bbe27a87e89fb6d55d356f0f

File: scripts/aws_replicate.py (hunk 24 above)
Comment (giangbui) [reply scripts/aws_replicate.py]: fixed in d47a36c889171c98bbe27a87e89fb6d55d356f0f

File: scripts/aws_replicate.py (hunk 35 above)
Comment (giangbui) [reply scripts/aws_replicate.py]: yeah. I also want to keep a local one

File: tests/test_replicate.py (hunk 38 above)
Comment (giangbui) [reply tests/test_replicate.py]: fixed in d47a36c889171c98bbe27a87e89fb6d55d356f0f

File: scripts/utils.py (hunk 46 above)
Comment (giangbui) [reply scripts/utils.py]: fixed in d47a36c889171c98bbe27a87e89fb6d55d356f0f

File: scripts/utils.py (hunk 47 above)
Comment (giangbui) [reply scripts/utils.py]: fixed in d47a36c889171c98bbe27a87e89fb6d55d356f0f

File: README.md (hunk 11 above)
Comment (giangbui) [reply README.md]: fixed in d47a36c889171c98bbe27a87e89fb6d55d356f0f

File: README.md (hunk 14 above)
Comment (giangbui) [reply README.md]: fixed in d47a36c889171c98bbe27a87e89fb6d55d356f0f

File: README.md (hunk 17 above)
Comment (giangbui) [reply README.md]: Google returns service account credential in json format

File: dataflow_pipeline.py (hunk 19 above)
Comment (giangbui) [reply dataflow_pipeline.py]: fixed in d47a36c889171c98bbe27a87e89fb6d55d356f0f

File: scripts/utils.py (hunk 41 above)
Comment (giangbui) [reply scripts/utils.py]: fixed in d47a36c889171c98bbe27a87e89fb6d55d356f0f

File: scripts/utils.py (hunk 42 above)
Comment (giangbui) [reply scripts/utils.py]: fixed in d47a36c889171c98bbe27a87e89fb6d55d356f0f

File: scripts/utils.py (hunk 43 above)
Comment (giangbui) [reply scripts/utils.py]: The function was removed since it is not used

File: scripts/utils.py (hunk 44 above)
Comment (giangbui) [reply scripts/utils.py]: remove it also

File: tests/test_replicate.py (hunk 45 above)
Comment (giangbui) [reply tests/test_replicate.py]: fixed in d47a36c889171c98bbe27a87e89fb6d55d356f0f

File: scripts/aws_replicate.py [hunk 49]
```diff
@@ -0,0 +1,236 @@
...
+                        )
+                        # should wait for safety
+                        subprocess.Popen(shlex.split(cmd + " --quiet")).wait()
+
+            if execstr != base_cmd:
+                subprocess.Popen(shlex.split(execstr + " --quiet")).wait()
+
+            index = index + chunk_size
+
+        global TOTAL_PROCESSED_FILES
```
Comment (rudyardrichter) [inline scripts/aws_replicate.py]: this isn't necessary anymore right?

File: scripts/aws_replicate.py [hunk 50]
```diff
@@ -0,0 +1,236 @@
...
+                    }
+                except Exception as e:
+                    logger.error(e.message)
+                    json_log[object_path] = {
+                        "copy_success": True,
+                        "index_success": False,
+                        "msg": e.message,
+                    }
+
+        global TOTAL_INDEXED_FILES
```
Comment (rudyardrichter) [inline scripts/aws_replicate.py]: same here, not needed now?

File: dataflow_pipeline.py (hunk 23 above)
Comment (rudyardrichter) [reply dataflow_pipeline.py]: +1 to this

File: scripts/aws_replicate.py (hunk 49 above)
Comment (giangbui) [reply scripts/aws_replicate.py]: just removed in caac0117d8f83af97684b250b3a50abfb033f05c

File: scripts/aws_replicate.py [hunk 51]
```diff
@@ -0,0 +1,230 @@
...
+                    }
+                except Exception as e:
+                    logger.error(e.message)
+                    json_log[object_path] = {
+                        "copy_success": True,
+                        "index_success": False,
+                        "msg": e.message,
+                    }
+
+        global total_indexed_files
```
Comment (rudyardrichter) [inline scripts/aws_replicate.py]: :point_right:

File: scripts/aws_replicate.py [hunk 52]
```diff
@@ -0,0 +1,230 @@
...
+        return existed_keys
+
+    def build_source_bucket_dataset(self):
+        """
+        build source bucket dataset for lookup
+        to avoid list object operations
+        """
+        s3 = boto3.resource("s3")
+        bucket = s3.Bucket(self.bucket)
+        return set([file.key for file in bucket.objects.all()])
```
Comment (rudyardrichter) [inline scripts/aws_replicate.py]:
```suggestion
         return {f.key for f in bucket.objects.all()}
```

File: dev-requirements.txt (hunk 26 above)
Comment (giangbui) [reply dev-requirements.txt]: We run on remote machines so we have to pack all dependencies. 
Indeed it is 1.5.6 but the there is error in setup.py of indexclient that makes the build as 1.5.2

File: scripts/replicate.py (hunk 39 above)
Comment (giangbui) [reply scripts/replicate.py]: Agree. It is sorta messy. Originally we want to support both google and aws replicate with multiple thread then we switch to google data flow. The repos should be divided into 2 repos, one for aws and one for google. I will spend time for refactor code in the next release

File: jobs/google-bucket-replicate-job.yaml (hunk 27 above)
Comment (giangbui) [reply jobs/google-bucket-replicate-job.yaml]: That is a bad example. You can use the same bucket or two different ones
